// webdriver-manager/built/lib/binaries/ie_driver.js
"use strict";
const config_1 = require("../config");
const binary_1 = require("./binary");
/**
 * The internet explorer binary.
 */
class IEDriver extends binary_1.Binary {
    constructor(alternateCDN) {
        super(alternateCDN || config_1.Config.cdnUrls().ie);
        this.name = 'IEDriverServer';
        this.versionCustom = IEDriver.versionDefault;
        this.prefixDefault = 'IEDriverServer';
        this.suffixDefault = '.zip';
        this.arch = config_1.Config.osArch();
    }
    id() {
        return IEDriver.id;
    }
    versionDefault() {
        return IEDriver.versionDefault;
    }
    version() {
        if (config_1.Config.osType() == 'Windows_NT') {
            if (this.arch == 'x64') {
                return '_x64_' + this.versionCustom;
            }
            else {
                return '_Win32_' + this.versionCustom;
            }
        }
        return '';
    }
    url() {
        let urlBase = this.cdn + this.shortVersion(this.versionCustom) + '/';
        let filename = this.prefix() + this.version() + this.suffix();
        return urlBase + filename;
    }
}
IEDriver.os = [binary_1.OS.Windows_NT];
IEDriver.id = 'ie';
IEDriver.versionDefault = config_1.Config.binaryVersions().ie;
IEDriver.isDefault32 = false;
IEDriver.isDefault64 = false;
exports.IEDriver = IEDriver;
