// require-npm/require.js
const path = require('path'),
    fs = require('fs'),
    processGenerator = require('child_process');

const Settings = {
    Library: path.join(process.env[(process.platform == 'win32') ? 'USERPROFILE' : 'HOME'], 'local-node-modules'),
    LibraryModulePath: "",
    ProjectModulePath: path.join(process.cwd(), 'node_modules')
};

Settings.LibraryModulePath = path.join(Settings.Library, 'node_modules');

module.exports = {
    decorate(require){

        require.global = GlobalRequire;
        require.local = LocalRequire;
        require.modules = ReadModules;

        return this;
    },
    settings: Settings
};

function ReadModules(){

    const modulePath = Settings.LibraryModulePath;
    return fs.readdirSync(modulePath).filter(fileName => fs.lstatSync(path.join(modulePath, fileName)).isDirectory());
}
function LocalRequire(){

    if(arguments.length === 0) return;

    var moduleList = Array.prototype.slice.apply(arguments);

    const importPathList = moduleList.
        map(moduleName => path.join(Settings.LibraryModulePath, moduleName));

    if(fs.existsSync(Settings.Library) === false)
        fs.mkdirSync(Settings.Library);

    if(fs.existsSync(Settings.ProjectModulePath) === false)
        fs.mkdirSync(Settings.ProjectModulePath);

    const json = path.join(Settings.Library, 'package.json'),
        exists = fs.existsSync(json);

    if(exists === false)
        processGenerator.execSync(`npm init --yes`, {cwd: Settings.Library});

    const installationList = importPathList.
        reduce(function(state, value, index){

            const importExists = fs.existsSync(value);

            if(importExists)
                return state;
            else
                return [...state, moduleList[index]];

        }, []);

    const installations = installationList
        .map(moduleName => {
            console.log(`..installing [${moduleName}]`);
            processGenerator.execSync(`npm install ${moduleName} --save-dev`, {cwd: Settings.Library});
            console.log('Finished.');
        });

    importPathList.forEach((importPath, index) => {

        const projectReference = path.join(Settings.ProjectModulePath, moduleList[index]);
        if(!fs.existsSync(projectReference))
        processGenerator.execSync(`ln -s ${importPath}`, {cwd: Settings.ProjectModulePath});
    });

    const importResultList = moduleList.map(moduleName => require(moduleName));

    return importResultList.length === 1? importResultList[0] : importResultList;
}

function GlobalRequire(){

    var argList = Array.prototype.slice.apply(arguments);

    const pathList = argList.map(arg => path.join('/usr/local/lib/node_modules', arg)),
        resultList = pathList.map(path => require(path));

    return resultList.length === 1? resultList[0] : resultList;
}
