// blocking-proxy/built/lib/config.js
#!/usr/bin/env node
"use strict";
var minimist = require('minimist');
var opts = {
    boolean: ['help', 'fork'],
    string: ['port', 'seleniumAddress'],
    alias: {
        help: ['h'],
        port: ['p'],
        seleniumAddress: ['s'],
    },
    default: {
        port: process.env.BP_PORT || 0,
        seleniumAddress: process.env.BP_SELENIUM_ADDRESS || 'http://localhost:4444/wd/hub',
        rootElement: 'body'
    }
};
function processArgs(argv) {
    return minimist(argv, opts);
}
exports.processArgs = processArgs;
function printHelp() {
    console.log("\nUsage: blocking-proxy <options>\n\nOptions:\n    --help, -h              Show help.\n    --port, -p              The port to listen on. If unset, will choose a random free port.\n    --fork                  Start in fork mode. BlockingProxy will use process.send() to communicate\n                                with the parent process.\n    --selenumAddress, -s    The address of the selenium remote server to proxy.\n    --rootElement           Element housing ng-app, if not html or body.\n");
}
exports.printHelp = printHelp;
